### ISSUE TITLE
Automatically generate Equatable definition for Swift

### ISSUE THREAD

**Issue by jasonsilberman** [apollographql/apollo-codegen#215]
I think it would be very helpful to automatically generate `Equatable` definitions for the Swift code that is generated in `API.swift`.

This would be very helpful because it would allow for comparison of fetch objects. I think if `GraphQLSelectionSet` was made to conform to `Equatable`, it would allow a lot of future use cases. One immediate use case is comparing objects to see if a `UITableView` needs to be updated or not. Without `GraphQLSelectionSet` conforming to `Equatable`, it is very hard to do this.

I don't think implementing this would be very hard, we could simply generate code that goes through each property and compares them.

**Comment by martijnwalraven** [comment]
Yep, I agree, that would definitely be useful. Maybe we could add in `Hashable` support as well. Want to take a stab at this?

**Comment by jasonsilberman** [comment]
Great! I think I might be able to that, but I am not really familiar with the codebase. Any hints to help me get started @martijnwalraven?

**Comment by martijnwalraven** [comment]
I think you'll probably want to take a look at [`structDeclarationForSelectionSet`](https://github.com/apollographql/apollo-codegen/blob/master/src/swift/codeGeneration.ts#L211), and add code to generate `Equable` (and possibly `Hashable`) support from `properties`.

**Comment by adysart** [comment]
Note that Swift 4 is supposedly bringing automatic synthesis of Equatable and Hashable for structs:

https://github.com/apple/swift-evolution/blob/master/proposals/0185-synthesize-equatable-hashable.md

So this might not be necessary in the long run.

**Comment by bclymer** [comment]
Swift 4.1 does bring this for free now, so I put up a PR - https://github.com/apollographql/apollo-cli/pull/578
